File: p740p758XML/p740XML/MyXMLParserDelegate.m


#import "MyXMLParserDelegate.h"


@implementation MyXMLParserDelegate
@synthesize text, parent, child, name;

- (void) start: (NSString*) el parent: (id) p {
    self.name = el;
    self.parent = p;
    self.text = [NSMutableString string];
}

- (void) makeChild: (Class) class elementName: (NSString*) elementName parser: (NSXMLParser*) parser {
    MyXMLParserDelegate* del = [[class alloc] init];
    self.child = del;
    parser.delegate = del;
    [del start: elementName parent: self];
    [del release];
}

- (void) finishedChild: (NSString*) s { // subclass implements as desired
}

- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string {
    [self.text appendString:string];
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName {
    if (parent) {
        [parent finishedChild: [[self.text copy] autorelease]];
        parser.delegate = self.parent;
    }
}

- (void) dealloc {
    [text release];
    [child release];
    [name release];
    [super dealloc];
}


@end

File: iOS10bookexamples/bk2ch23p818xml/ch36p1070xml/MyXMLParserDelegate.swift


import UIKit


class MyXMLParserDelegate : NSObject {
    var name : String!
    var text = ""
    weak var parent : MyXMLParserDelegate?
    var child : MyXMLParserDelegate!
    
    required init(name:String, parent:MyXMLParserDelegate?) {
        self.name = name
        self.parent = parent
        super.init()
    }
    
    func makeChild(_ klass:MyXMLParserDelegate.Type, elementName:String, parser:XMLParser) {
        let del = klass.init(name:elementName, parent:self)
        self.child = del
        parser.delegate = del
    }
    func finishedChild(_ s:String) {
        fatalError("Subclass must implement finishedChild:!")
    }
    
}

extension MyXMLParserDelegate : XMLParserDelegate {
    func parser(_ parser: XMLParser, foundCharacters string: String) {
        self.text = self.text + string
    }
    
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if self.parent != nil {
            self.parent!.finishedChild(self.text)
            parser.delegate = self.parent
        }
    }
    
}


